model/histogram/float_histogram.go

```diff
@@ -316,6 +317,30 @@ func (h *FloatHistogram) Equals(h2 *FloatHistogram) bool {
 	return true
 }
 
+// Size returns the size of the whole fields histogram in bytes.
+func (fh *FloatHistogram) Size() int {
```

Comment:
I'm pretty sure we cannot have this in the hot path. It's way too expensive. Also, we should avoid the `unsafe` package wherever we can.

This is great to learn about the size, but once we know it, we should hardcode all the numbers currently retrieved via `unsafe.Sizeof`. This can be a bit tricky because some of them might depend on the architecture. If you find a clever way to take that into account, great. But we can also assume 64bit architectures for the purpose of this safeguard. (Prometheus on 32bit isn't really viable for larger payloads anyway.

Reply:
> I'm pretty sure we cannot have this in the hot path.

Do you mean as part of the `FloatHistogram`? Why not? Where would you put it instead?

> It's way too expensive. 

Expensive in terms of computation? Is it really?

> Also, we should avoid the `unsafe` package wherever we can.

Why? It's been used in other parts of the codebase

Reply:
> > I'm pretty sure we cannot have this in the hot path.
> 
> Do you mean as method of the `FloatHistogram` struct? Why not? Where would you put it instead?

I guess it's fine to have a "size" method, but we cannot calculate the size from first principles each time. Most of it is static and only depends on the architecture. The dynamic parts (i.e. `len` of slices) is OK to calculate dynamically each time. (At least we can try and see if it has an unacceptable impact on performance.)

> > It's way too expensive. 
>
> Expensive in terms of computation? Is it really?

This is just housekeeping. It shouldn't take more than a few percent of total compute time. Plus, it's easy to avoid all the "static" parts. As said above, just the various `len` of slices need to be calculated dynamically. And that's (maybe and hopefully) cheap enaugh.

> > Also, we should avoid the `unsafe` package wherever we can.
>
> Why? It's been used in other parts of the codebase

And it's _very_ frowned upon. (Anecdote: At Gophercon, one of the semi-famous Gophers there asked my what I'm working on. I said: "Prometheus". They said: "I feel sorry for you." I: "Why?" They: "Isn't it using the `unsafe` package?") Obviously, we shouldn't let anecdotes drive our development, but any use of `unsafe` needs an _exceptionally_ strong justification because it breaks the safety guarantees of Go. Even the current uses are often doubted (also within the Prometheus community, not only by outsiders). And this use here is very easy to avoid, see above.

So I would suggest to pre-calculate all the "static" parts (how large is the baseline usage of a slice, how large is a float64, etc.) and just put the constant values in the actual code (with a comment how you calculated it), and then add the dynamic parts (e.g. `len(fh.PositiveSpans) * 8`, assuming a `Span` takes 8 bytes) in a dynamic fashion. You can just do it for 64bit architectures for now and add a TODO to add refined calculation for 32bit architectures later (or do it right now if you like, you can use suffixes in the file name or build tags to decide which code is used for which architecture).